Extended keys serialization and path generation / derivation

A-Zak opened the pull request:
Hey Kevin,

I've added the extended keys serialization and also methods to generate key paths and to derive child keys with a path. I've also added some tests - for the serialization I've tested according to the BIP32 test vectors, and for the path, I've just tried to cover all edge cases...
I also tried to comply to your style :D

Cheers...

A-Zak committed:
Adding vesion and parent pointer to ExtendedECKey

A-Zak committed:
sprinkle in some tests

A-Zak committed:
add the needed properties for the extended keys

A-Zak committed:
fix parent fingerprint

A-Zak committed:
add extended keys serilization + tests

A-Zak committed:
fix indention to match the rest of the repo

A-Zak committed:
Merge branch 'feature/extended-keys' into develop

A-Zak committed:
Refactor method names to conform with BIP32

and sprinkle some comments

A-Zak committed:
Add path generation & deriviation + tests

A-Zak committed:
fix silly mistake

never write crypto drunk... :)

A-Zak committed:
Move ExtendedKeyVersion into BitcoinParams

kgreenek commented:
Wow thanks! I'll try to take a look at this tonight.

One piece of feedback I would give after taking a glance is to take a look
at the "params" pattern I've been starting to move to. For example see how
mainnetparameters conforms to the addressparameters protocol. This makes
things more flexible in the future, for example for alt coins.

On Monday, March 23, 2015, Alex Zak <notifications@github.com> wrote:

> Hey Kevin,
>
> I've added the extended keys serialization and also methods to generate
> key paths and to derive child keys with a path. I've also added some tests
> - for the serialization I've tested according to the BIP32 test vectors,
> and for the path, I've just tried to cover all edge cases...
> I also tried to comply to your style :D
>
> Cheers...
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/DoubleSha/BitcoinSwift/pull/10
> Commit Summary
>
>    - Adding vesion and parent pointer to ExtendedECKey
>    - sprinkle in some tests
>    - add the needed properties for the extended keys
>    - fix parent fingerprint
>    - add extended keys serilization + tests
>    - fix indention to match the rest of the repo
>    - Merge branch 'feature/extended-keys' into develop
>    - Refactor method names to conform with BIP32
>    - Add path generation & deriviation + tests
>
> File Changes
>
>    - *M* BitcoinSwift/ExtendedECKey.swift
>    <https://github.com/DoubleSha/BitcoinSwift/pull/10/files#diff-0> (210)
>    - *M* BitcoinSwiftTests/ExtendedECKeyTests.swift
>    <https://github.com/DoubleSha/BitcoinSwift/pull/10/files#diff-1> (127)
>
> Patch Links:
>
>    - https://github.com/DoubleSha/BitcoinSwift/pull/10.patch
>    - https://github.com/DoubleSha/BitcoinSwift/pull/10.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/DoubleSha/BitcoinSwift/pull/10>.
>

A-Zak commented:
I've just refactored that part erlier today :)

On Tue, Mar 24, 2015 at 7:30 PM Kevin Greene <notifications@github.com>
wrote:

> Wow thanks! I'll try to take a look at this tonight.
>
> One piece of feedback I would give after taking a glance is to take a look
> at the "params" pattern I've been starting to move to. For example see how
> mainnetparameters conforms to the addressparameters protocol. This makes
> things more flexible in the future, for example for alt coins.
>
> On Monday, March 23, 2015, Alex Zak <notifications@github.com> wrote:
>
> > Hey Kevin,
> >
> > I've added the extended keys serialization and also methods to generate
> > key paths and to derive child keys with a path. I've also added some
> tests
> > - for the serialization I've tested according to the BIP32 test vectors,
> > and for the path, I've just tried to cover all edge cases...
> > I also tried to comply to your style :D
> >
> > Cheers...
> > ------------------------------
> > You can view, comment on, or merge this pull request online at:
> >
> > https://github.com/DoubleSha/BitcoinSwift/pull/10
> > Commit Summary
> >
> > - Adding vesion and parent pointer to ExtendedECKey
> > - sprinkle in some tests
> > - add the needed properties for the extended keys
> > - fix parent fingerprint
> > - add extended keys serilization + tests
> > - fix indention to match the rest of the repo
> > - Merge branch 'feature/extended-keys' into develop
> > - Refactor method names to conform with BIP32
> > - Add path generation & deriviation + tests
> >
> > File Changes
> >
> > - *M* BitcoinSwift/ExtendedECKey.swift
> > <https://github.com/DoubleSha/BitcoinSwift/pull/10/files#diff-0> (210)
> > - *M* BitcoinSwiftTests/ExtendedECKeyTests.swift
> > <https://github.com/DoubleSha/BitcoinSwift/pull/10/files#diff-1> (127)
> >
> > Patch Links:
> >
> > - https://github.com/DoubleSha/BitcoinSwift/pull/10.patch
> > - https://github.com/DoubleSha/BitcoinSwift/pull/10.diff
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/DoubleSha/BitcoinSwift/pull/10>.
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/DoubleSha/BitcoinSwift/pull/10#issuecomment-85610002>.
>

kgreenek commented:
Sorry for the delay. Taking a look now. In general it looks good. Please watch your style - especially extra whitespace. I'm sorry there isn't better documentation; up until now it's been mostly just me working on this, so I'll try to point you to what I'm thinking.

kgreenek commented on BitcoinSwift/BitcoinMainNetParameters.swift:
```diff
@@ -49,4 +49,14 @@ public class BitcoinMainNetParameters: BitcoinParameters {
+  
+  // MARK: - ExtendedKeyVersionParameters
+  
+  public var addresses: (pub:UInt32, prv: UInt32) {
```
The name "addresses" is very unclear. These are not addresses. Also remember that the name will be part of the BitcoinMainNetParameters class (for example), so the names need to be unique. This is true of protocols in general. You should make the name sufficiently unique that somebody implementing the protocol won't have a name collision with another protocol implemented by the same class.

I recommend splitting this into two separate vars like:
extendedPublicKeyVersion & extendedPrivateKeyVersion

Also in general, names should be fully typed out. E.g. instead of "pub" just type out "publicKey". The whole rest of the code base is this way.

kgreenek commented on BitcoinSwift/ExtendedECKey.swift:
```diff
@@ -8,18 +8,144 @@
 
 import Foundation
 
+
```
nit: extra whitespace.

kgreenek commented on BitcoinSwift/ExtendedECKey.swift:
```diff
@@ -8,18 +8,144 @@
+
+
+
+public enum KeyType {
```
I don't think KeyType is really an accurate way to represent this. An ECKey can have either:
- both a private key and a public key
or
- just a public key

In the ECKey class, this is already represented by checking whether the private key is nil. So this is really a key pair, where the private key can be nil in some cases.

kgreenek commented on BitcoinSwift/ExtendedECKey.swift:
```diff
@@ -8,18 +8,144 @@
+  public let parent: ExtendedECKey?
+  
+  
+  // This key's identifier (this corresponds exactly to a traditional bitcoin address)
```
nit: comments should always end with a period (sorry, I know there isn't a published style guide. Please bear with all my nits.)

kgreenek commented on BitcoinSwift/ExtendedECKey.swift:
```diff
@@ -8,18 +8,144 @@
+  
+  
+  // This key's identifier (this corresponds exactly to a traditional bitcoin address)
+  public var identifier:NSData {
```
nit: space after the colon

kgreenek commented on BitcoinSwift/ExtendedECKey.swift:
```diff
@@ -8,18 +8,144 @@
+  }
+  
+  // Return this key's fingerprint
+  public var fingerprint:NSData {
```
nit: here too. and elsewhere in this PR. There should be a space after the colon.

kgreenek commented on BitcoinSwift/ExtendedECKey.swift:
```diff
@@ -8,18 +8,144 @@
+    return publicKey.SHA256Hash().RIPEMD160Hash()
+  }
+  
+  // Return this key's fingerprint
```
nit: Period after comments.

kgreenek commented on BitcoinSwift/ExtendedECKey.swift:
```diff
@@ -8,18 +8,144 @@
+      return parent.fingerprint
+    }
+    else {
+      let masterprint: [UInt8] = [
```
Define masterFingerprint as a class var. Also, don't shorten names. Just call it masterFingerprint.

kgreenek commented on BitcoinSwift/ExtendedECKey.swift:
```diff
@@ -8,18 +8,144 @@
+    }
+  }
+  
+  // Find the depth of this key
```
Here and elsewhere. If you use "///" instead of just "//" when commenting functions and classes, then when the user control+clicks on the function, they'll see the documentation in the little bubble that pops up. Similar to Apple's framework classes.

Please do that on the other functions/vars in this PR too.

kgreenek commented on BitcoinSwift/ExtendedECKey.swift:
```diff
@@ -8,18 +8,144 @@
+  
+  // True if this key has no parent
+  public var isMaster:Bool {
+    return self.parent == nil
```
nit: no need for "self."

kgreenek commented on BitcoinSwift/ExtendedECKey.swift:
```diff
@@ -8,18 +8,144 @@
+  public var master:ExtendedECKey {
+    if let parent = self.parent {
+      return parent.master
+    }
```
nit: the "else" should be on the same line as the "}". Elsewhere in this PR too.

kgreenek commented on BitcoinSwift/ExtendedECKey.swift:
```diff
@@ -8,18 +8,144 @@
+  
+  // Serialize the extended key data
+  public func serializeExtendedKey(ofType type:KeyType = .PublicKey, version: ExtendedKeyVersionParameters = BitcoinMainNetParameters.get()) -> NSMutableData {
+    
```
nit: extra whitespace.

kgreenek commented on BitcoinSwift/ExtendedECKey.swift:
```diff
@@ -8,18 +8,144 @@
+  public func serializeExtendedKey(ofType type:KeyType = .PublicKey, version: ExtendedKeyVersionParameters = BitcoinMainNetParameters.get()) -> NSMutableData {
+    
+    func keyDataForType(type:KeyType) -> NSData {
+      
```
extra whitespace here too.

kgreenek commented on BitcoinSwift/ExtendedECKey.swift:
```diff
@@ -8,18 +8,144 @@
+  }
+  
+  // Serialize the extended key data
+  public func serializeExtendedKey(ofType type:KeyType = .PublicKey, version: ExtendedKeyVersionParameters = BitcoinMainNetParameters.get()) -> NSMutableData {
```
Remove the "KeyType" enum and just have two separate functions "serializedExtendedPublicKey" and "serializedExtendedPrivateKey".

kgreenek commented on BitcoinSwift/ExtendedECKey.swift:
```diff
@@ -8,18 +8,144 @@
+  }
+  
+  // Encode the extended key into a base64check string
+  public func encodeExtendedKey(ofType type:KeyType = .PublicKey, version: ExtendedKeyVersionParameters = BitcoinMainNetParameters.get()) -> String {
```
Sorry, I know this isn't documented anywhere, the the maximum line length is 100 chars. Please wrap when longer than that.

kgreenek commented on BitcoinSwift/ExtendedECKey.swift:
```diff
@@ -81,13 +207,64 @@ public class ExtendedECKey : ECKey {
 
+  // Derive a key from the path.
+  // Supports both absolute or relative paths.
+  public func derive(path:String, isAbsolute isAbsolutePath:Bool = true) -> ExtendedECKey? {
```
I would call this deriveFromExtendedKeyPath. Please be explicit with function/variable names.

kgreenek commented on BitcoinSwift/ExtendedECKey.swift:
```diff
@@ -81,13 +207,64 @@ public class ExtendedECKey : ECKey {
   }
 
+  // Derive a key from the path.
+  // Supports both absolute or relative paths.
```
Add comments here about the format of the extended key path.

kgreenek commented on BitcoinSwiftTests/ExtendedECKeyTests.swift:
```diff
@@ -287,4 +287,131 @@ class ExtendedECKeyTests: XCTestCase {
+    XCTAssert(tGrandChild.version === BitcoinTestNetParameters.get())
+  }
+  
+  func testExtendedKeySerilization_TestVector1() {
```
nit: No need for the underscore. The codebase doesn't do that anywhere else.

kgreenek commented on BitcoinSwiftTests/ExtendedECKeyTests.swift:
```diff
@@ -287,4 +287,131 @@ class ExtendedECKeyTests: XCTestCase {
+    XCTAssertEqual(key.encodeExtendedKey(ofType: .PublicKey),  "xpub68Gmy5EdvgibQVfPdqkBBCHxA5htiqg55crXYuXoQRKfDBFA1WEjWgP6LHhwBZeNK1VTsfTFUHCdrfp1bgwQ9xv5ski8PX9rL2dZXvgGDnw")
+    XCTAssertEqual(key.encodeExtendedKey(ofType: .PrivateKey), "xprv9uHRZZhk6KAJC1avXpDAp4MDc3sQKNxDiPvvkX8Br5ngLNv1TxvUxt4cV1rGL5hj6KCesnDYUhd7oWgT11eZG7XnxHrnYeSvkzY7d2bhkJ7")
+    
+    // chain - m/0ʜ/1
```
These tests are awesome, thank you!

kgreenek commented on BitcoinSwift/ExtendedECKey.swift:
```diff
@@ -36,7 +162,7 @@ public class ExtendedECKey : ECKey {
   /// Can return nil in the (very very very very) unlikely case the randomly generated private key
   /// is invalid. If nil is returned, retry.
-  public class func masterKeyWithSeed(seed: SecureData) -> ExtendedECKey? {
+  public class func masterKeyWithSeed(seed: SecureData, version:BitcoinParameters = BitcoinMainNetParameters.get()) -> ExtendedECKey? {
```
Call this "params" instead of version.

A-Zak committed:
comfront to styleguide

A-Zak committed:
explicifation refactoring

A-Zak committed:
more refactoring

reomve the KeyType enum - instead seporated into public / private functions.
rename version to params
and remove the silly named addresses var and repalced it with two functions instead.

A-Zak committed:
Add some comments

& make function comments triple slashed for nicer xcode commenting...

A-Zak committed:
change the path seporator to conform with Bip32

A-Zak committed:
simplify init method signatures...

A-Zak committed:
remove some empty lines

A-Zak commented:
Hey,

First of all, sorry it took me so long... We had some very tight schedules and I had very little time to spare...

About the comments, I can appreciate the style guide, and I've changed most of the things you've asked - all but two comments:
1) the max line length (100 chars) - I think it sometimes looks very bad when a line of code is broken down to several lines, especially when combined with the "no empty lines under any circumstance" style guide. So I left a couple of more than 100 char lines where I thought it looks better...
and 2) the default value for the params parameter in the initializers. I understand that creating a dependency to BitcoinMainNetParams makes it a tiny bit harder for someone to reuse this code for other alt coins, but I think that it's more than understandable that **Bitcoin**Swift will have a dependencies to Bitcoin classes, not to mention the great care you took to make it extensively 'protocoled' to make any such changes extremely easy...
In my opinion most of the code using BitcoinSwift will be used to on the main net, and with the default value in place, the benefits of a cleaner and more maintainable code, which is also less error prone, are far greater.

Let me know what you think... :)

A-Zak commented:
P.S.
There's another pull request waiting with xCode 6.3 / Swift 1.2 compatibility fixes...

kgreenek commented:
No worries. This is just a part time thing for me too :)

1) The char-limit is a hard limit. Please wrap everything so it's all under 100 chars. There is almost always a way to wrap it that looks nice, but in the rare cases when there isn't, it's still better to wrap manually than have your editor do the line-wrapping or scroll.

2) I'm worried more about people accidentally creating something for mainnet when they meant testnet and having a mismatch that breaks things in difficult ways. I always want to require params to be explicitly passed to it's clear what the intention is. Also, it should follow the pattern of passing params as the first argument.

Another thing, eventually I'm going to split up the codebase into libraries, and I don't want to introduce a dependency on the "BitcoinParams" library (or whatever it ends up being) when there shouldn't be. I'm a big supporter of keeping things 100% semantically pure and modular.

kgreenek commented on BitcoinSwift/ExtendedECKey.swift:
```diff
@@ -8,18 +8,151 @@
+                 length: Static.masterParentFingerprint.count)
+  }
+  
+  private class func pathSeporator() -> Character {
```
nit: Spelling

kgreenek commented on BitcoinSwift/ExtendedECKey.swift:
```diff
@@ -8,18 +8,151 @@
 /// BIP 32: https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki
 public class ExtendedECKey : ECKey {
-
+  
```
nit: trailing whitespace

kgreenek commented on BitcoinSwift/BitcoinTestNetParamaters.swift:
```diff
@@ -49,4 +49,15 @@ public class BitcoinTestNetParameters: BitcoinParameters {
+  public var extendedPublicKeyVersion: NSData {
+    let publicKeyVersion: [UInt8] = [0x04, 0x35, 0x87, 0xcf]
+    return NSData(bytes: publicKeyVersion, length: publicKeyVersion.count)
+  }
```
nit: add a nl

kgreenek commented on BitcoinSwift/BitcoinMainNetParameters.swift:
```diff
@@ -49,4 +49,15 @@ public class BitcoinMainNetParameters: BitcoinParameters {
+  public var extendedPublicKeyVersion: NSData {
+    let publicKeyVersion: [UInt8] = [0x04, 0x88, 0xb2, 0x1e]
+    return NSData(bytes: publicKeyVersion, length: publicKeyVersion.count)
+  }
```
nit: add a nl

kgreenek commented on BitcoinSwift/ExtendedECKey.swift:
```diff
@@ -8,18 +8,151 @@
+  var extendedPrivateKeyVersion: NSData { get }
+}
+
+
```
nit: extra whitespace

kgreenek commented on BitcoinSwift/ExtendedECKey.swift:
```diff
@@ -8,18 +8,151 @@
+  public func serializedExtendedPrivateKey(params: ExtendedKeyVersionParameters = BitcoinMainNetParameters.get()) -> NSMutableData {
+    let keyData = NSMutableData()
+    keyData.appendUInt8(0x00)
+    keyData.appendData(self.privateKey.mutableData)
```
Whenever you are storing a private key in memory, you always need to be using SecureData.

kgreenek commented on BitcoinSwift/ExtendedECKey.swift:
```diff
@@ -81,13 +214,67 @@ public class ExtendedECKey : ECKey {
     assert(Int32(childPrivateKey.length) == ECKey.privateKeyLength())
     let childChainCode = indexHash[32..<64]
-    return ExtendedECKey(privateKey: childPrivateKey, chainCode: childChainCode, index: index)
+    return ExtendedECKey(privateKey: childPrivateKey, chainCode: childChainCode, index: index, parent:self)
```
nit: space after colon

kgreenek commented on BitcoinSwift/ExtendedECKey.swift:
```diff
@@ -8,18 +8,151 @@
+    return seporator.join(self.pathComponents)
+  }
+  
+  private func serializedExtendedKey(#version: NSData, keyData: NSData) -> NSMutableData {
```
What's the # in front of version?

kgreenek replied:
Also, this should be called serializedExtendedKeyWithVersion

kgreenek commented:
Just following up with this. It's really close. I really do want this to be merged. I may pull it in and make the remaining changes myself too if you're busy :)
